src/Tooltip/Tooltip.tsx

```diff
@@ -8,37 +11,64 @@ export interface TooltipMessageGroupItem {
     message: string;
 }
 
-interface TooltipState {
-    toggle: boolean;
-}
-
+export type TooltipTrigger = "hover" | "click" | "focus";
 export type TooltipTheme = "default" | "light" | "primary" | "warning" | "success" | "danger" | "purple";
-export type TooltipPosition = "top" | "bottom" | "left" | "right" | "top-right" | "top-left" | "bottom-right" | "bottom-left" | "left-top" | "left-bottom" | "right-top" | "right-bottom";
+export type TooltipPosition = ElementPosition;
 
 export interface TooltipProps {
     className?: string;
     customSvg?: React.ReactNode;
     id?: string;
+    ref?: React.LegacyRef<Tooltip>;
+    /** @deprecated use content instead */
     message?: string;
+    /** @deprecated use content instead */
     messageGroup?: Array<TooltipMessageGroupItem>;
-    onClick?: (event?: React.MouseEvent<HTMLDivElement>) => void;
+    /** @deprecated use onVisibleChange instead */
+    onClick?: (event?: React.MouseEvent<HTMLDivElement> | React.FocusEvent<HTMLElement> | React.TouchEvent<HTMLDivElement>) => void;
     position?: TooltipPosition;
     theme?: TooltipTheme;
+    /** @deprecated use content instead */
     title?: string;
+    /** @deprecated use trigger instead */
     triggerOnHover?: boolean;
+    /** @deprecated */
     width?: number;
+    content?: string | React.ReactNode;
+    trigger?: TooltipTrigger;
+    disableAutoPosition?: boolean;
+    onVisibleChange?: (event: React.MouseEvent<HTMLDivElement> | React.FocusEvent<HTMLElement> | React.TouchEvent<HTMLDivElement>, visible: boolean) => void;
+}
+interface TooltipState {
+    visible: boolean;
+    referenceId: string;
+    tooltipContainer: HTMLDivElement;
 }
 
 export class Tooltip extends React.Component<TooltipProps, TooltipState> {
+    private containerRef: React.RefObject<HTMLDivElement> = React.createRef();
+
     constructor(props: TooltipProps) {
         super(props);
 
         this.state = {
-            toggle: false
+            visible: false,
+            referenceId: "",
```

Comment:
Any particular reason why the referenceId is being set on componentDidMount and not here?

Reply:
removed didmount, inititate in state